source/target/realtek/rtl8195am/target.c

```diff
@@ -0,0 +1,97 @@
+/**
+ * @file    target.c
+ * @brief   Target information for Realtek RTL8195AM
+ *
+ * DAPLink Interface Firmware
+ * Copyright (c) 2016-2017, Realtek Semiconductor Corp., All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "target_config.h"
+
+// The file flash_blob.c must only be included in target.c
+#include "flash_blob.c"
+
+// target information
+target_cfg_t target_device = {
+    .sector_size    = KB(4),
+    .sector_cnt     = MB(2) / KB(4),
+    .flash_start    = 0x00000000,
+    .flash_end      = 0x00000000 + MB(2),
+    .ram_start      = 0x10007000,
+    .ram_end        = 0x10070000,
+    .flash_algo     = (program_target_t *) &flash,
+    .erase_reset    = 0,
+};
+
+// RTL8195AM's main cpu can only talk 38400 with DAP UART
+//#include "string.h"
+#include "RTL.h"
+#include "rl_usb.h"
+#include "usb_for_lib.h"
+
+extern int32_t data_send_access;
+extern int32_t data_send_active;
+extern int32_t data_send_zlp;
+extern int32_t data_to_send_wr;
+extern int32_t data_to_send_rd;
+extern uint8_t *ptr_data_to_send;
+extern uint8_t *ptr_data_sent;
+extern int32_t data_read_access;
+extern int32_t data_receive_int_access;
+extern int32_t data_received_pending_pckts;
+extern int32_t data_no_space_for_receive;
+extern uint8_t *ptr_data_received;
+extern uint8_t *ptr_data_read;
+extern uint16_t control_line_state;
+extern CDC_LINE_CODING line_coding;
+
+extern int32_t USBD_CDC_ACM_PortReset(void);
+extern int32_t USBD_CDC_ACM_PortSetLineCoding(CDC_LINE_CODING *line_coding);
+
```

Comment:
@M-ichae-l  I am thinking to remove these externs.

Method 1
I think overriding the USBD_CDC_ACM_SetLineCoding is enough to communicate to the target.

In target.c
```
#include "uart.h"
 static UART_Configuration UART_Config;
int32_t USBD_CDC_ACM_SetLineCoding(void)
{
    UART_Config.Baudrate    = 38400;
    UART_Config.DataBits    = UART_DATA_BITS_8;
    UART_Config.Parity      = UART_PARITY_NONE;
    UART_Config.StopBits    = UART_STOP_BITS_1;
    UART_Config.FlowControl = UART_FLOW_CONTROL_NONE;
    return uart_set_configuration(&UART_Config);
 
}
```

Method 2
 You may have to add a weak function in file usbd_cdc_acm.c

Replace USBD_CDC_ACM_Reset function by lines of codes below:
```
//override for line_coding settings and return 1
__weak int32_t USBD_CDC_ACM_TargetLineCoding(CDC_LINE_CODING *line_coding)
{
    return (0);
}

/** \brief  Reset of the USB CDC class (ACM) variables and states

    The function resets class variables and states, it calls
    USBD_CDC_ACM_PortReset function which resets Virtual COM Port variables
    and states and calls USBD_CDC_ACM_PortSetLineCoding function with
    default parameters to set default communication settings for the
    Virtual COM Port.

    \return             0        Function failed.
    \return             1        Function succeeded.
 */


__weak int32_t USBD_CDC_ACM_Reset(void)
{
    data_send_access            = 0;
    data_send_active            = 0;
    data_send_zlp               = 0;
    data_to_send_wr             = 0;
    data_to_send_rd             = 0;
    ptr_data_to_send            = USBD_CDC_ACM_SendBuf;
    ptr_data_sent               = USBD_CDC_ACM_SendBuf;
    data_read_access            = 0;
    data_receive_int_access     = 0;
    data_received_pending_pckts = 0;
    data_no_space_for_receive   = 0;
    ptr_data_received           = USBD_CDC_ACM_ReceiveBuf;
    ptr_data_read               = USBD_CDC_ACM_ReceiveBuf;
    control_line_state          = 0;
    USBD_CDC_ACM_PortReset();

   //Check for default line coding for target
    if(USBD_CDC_ACM_TargetLineCoding(&line_coding) == 0){
        line_coding.dwDTERate       = 9600;
        line_coding.bCharFormat     = 0;
        line_coding.bParityType     = 0;
        line_coding.bDataBits       = 8;
    }
    
    return (USBD_CDC_ACM_PortSetLineCoding(&line_coding));
}
```
Then in your target.c, you could implement
```
#include "RTL.h"
#include "rl_usb.h"
#include "uart.h"

static CDC_LINE_CODING target_line = {
    .dwDTERate       = 38400,
    .bCharFormat     = 0,
    .bParityType     = 0,
    .bDataBits       = 8
};

static UART_Configuration UART_Config;

int32_t USBD_CDC_ACM_TargetLineCoding(CDC_LINE_CODING *line_coding)
{
    line_coding->dwDTERate       = target_line.dwDTERate;
    line_coding->bCharFormat     = target_line.bCharFormat;
    line_coding->bParityType     = target_line.bParityType;
    line_coding->bDataBits       = target_line.bDataBits;
    return (1);
}


int32_t USBD_CDC_ACM_SetLineCoding(void)
{

    UART_Config.Baudrate    = target_line.dwDTERate;
    UART_Config.DataBits    = (UART_DataBits) target_line.bDataBits;
    UART_Config.Parity      = (UART_Parity)   target_line.bParityType;
    UART_Config.StopBits    = (UART_StopBits) target_line.bCharFormat;
    UART_Config.FlowControl = UART_FLOW_CONTROL_NONE;
    return uart_set_configuration(&UART_Config);
 
}
```
But I think method 1 should be enough as all usb2uart will start with this handshake, let me know if I am wrong.